### PR TITLE
feat: one step product creation

### LINKED ISSUE #1508: One step to create product rather than 2 steps
Lots user think its finished after they click the "create product" button, on my website, user will just left the page after they click the "create product", they think its done. they just miss everything on the second "Edit Product" page, like completely miss the part of create variable product. and shipping things..
And the second page which it shouldn't call "edit product", because its still in the process of "creating product" for variable, shipping and lots other things.

My suggestion:
1: combine 2 steps into just one step page to create the product, no need for the "edit product" page at all, which will make things much more clear for users.
2: OR, move attribute/variable, shipping these basic things into the first step, otherwise vendors just missing these basic part after they click the "create product"

### PR DESCRIPTION
**Pull request by shohag121**
### All Submissions:

* [x] My code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)
* [x] My code satisfies feature requirements
* [ ] My code is tested
* [ ] My code passes the PHPCS tests
* [ ] My code has proper inline documentation
* [ ] I've included developer documentation (optional)
* [ ] I've added proper labels to this pull request

### Changes proposed in this Pull Request:

* Added One step product creation-related settings
*  Based on Admin settings product status select field changed to **Online** or **Pending Review**
* One-page/single-step product addition added. this closes #1508

Screenshot 2023-08-14 at 10 19 18 AM

image

### How to test the changes in this Pull Request:

1. Please check issue #1508
1. related https://github.com/weDevsOfficial/dokan-pro/pull/2434

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

**feat**: One Step product addition
Previously, vendors had to use a product popup or a limited field page to create a product. Now Dokan is giving one step single-page product addition experience.

### PR Self Review Checklist:

* Code is not following code style guidelines
* Bad naming: make sure you would understand your code if you read it a few months from now.
* KISS: Keep it simple, Sweetie (not stupid!).
* DRY: Don't Repeat Yourself.
* Code that is not readable: too many nested 'if's are a bad sign.
* Performance issues
* Complicated constructions that need refactoring or comments: code should almost always be self-explanatory.
* Grammar errors.

### FOR PR REVIEWER ONLY:

> As a reviewer, your feedback should be focused on the idea, not the person. Seek to understand, be respectful, and focus on constructive dialog.

> As a contributor, your responsibility is to learn from suggestions and iterate your pull request should it be needed based on feedback. Seek to collaborate and produce the best possible contribution to the greater whole.

* [ ] Correct — Does the change do what it’s supposed to? ie: code 100% fulfilling the requirements?
* [ ] Secure — Would a nefarious party find some way to exploit this change? ie: everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities?
* [ ] Readable — Will your future self be able to understand this change months down the road?
* [ ] Elegant — Does the change fit aesthetically within the overall style and architecture?

### CODE REVIEW

**Comment by nurul-umbhiya** [comment]
Related Issue: https://github.com/weDevsOfficial/dokan-pro/issues/1234

File: assets/src/js/product-editor.js [hunk 1]
```diff
@@ -70,9 +70,44 @@
...
             $( 'body' ).trigger( 'dokan-product-editor-loaded', this );
         },
 
+        setCorrectProductId : function () {
+            let productForm = $('.dokan-product-edit-form');
+            if ( ! productForm ) {
+              return;
+            }
+            let productId    = $( '#dokan_product_id' ).val();
+            let dokanOnePage = '1' === $('#dokan_one_page').val();
```
Comment (nurul-umbhiya) [inline assets/src/js/product-editor.js:84]: this can be obtain from i18n data

File: assets/src/js/product-editor.js [hunk 2]
```diff
@@ -70,9 +70,44 @@
...
+                // change the search property of the main url
+                url.search = searchParams.toString();
+
+                // the new url string
+                let newUrl = url.toString();
+                let stateData = {
+                    product_id: productId,
+                }
+
+                if ( 'edit' === action && dokanOnePage ) {
```
Comment (nurul-umbhiya) [inline assets/src/js/product-editor.js:105]: we can early return if dokanOnePage is not enabled, Also, dokanOnePage doesn't make sense, kindly consider changing this variable.

File: templates/products/edit-product-single.php [hunk 3]
```diff
@@ -30,21 +31,38 @@
...
     $post_id        = intval( $_GET['product_id'] );
     $post           = get_post( $post_id );
-    $post_title     = $post->post_title;
-    $post_content   = $post->post_content;
-    $post_excerpt   = $post->post_excerpt;
     $post_status    = $post->post_status;
+    $auto_draft     = 'auto-draft' === $post_status;
+    $post_title     = $auto_draft ? '' : $post->post_title;
+    $post_content   = $auto_draft ? '' : $post->post_content;
+    $post_excerpt   = $post->post_excerpt;
```
Comment (nurul-umbhiya) [inline templates/products/edit-product-single.php:38]: wouldn't it be better if we created auto-draft product via js and redirect to product edit page?
we can determine if this is a new product or not via post_status = auto-draft?

**Review by nurul-umbhiya — CHANGES_REQUESTED** [review]

**Comment by shohag121** [comment]
Tareq Bhaya instructed us to remove the admin settings related to product creation.

**Comment by tanjiralmamun** [comment]
> Tareq Bhaya instructed us to remove the admin settings related to product creation.

@shohag121 Bhai, many users still like and want the use the popup form. Also, many users are already used to it. So, removing an existing feature without prior notice or removing the popup form completely would not be a good decision in my opinion.

**Comment by shohag121** [comment]
@tanjiralmamun Bhai,

Initially, The feature had the support to choose the style of the product creation page. Tareq Bhaya instructed us to remove the settings after seeing it.

@nurul-umbhiya Bhai Already ran a user opinion poll in our FB group discussing it.

You are welcome to start a new discussion with @shams-sadek1981 Bhai and @Mohaiminulislam1989 Bhai. That would be best.

**Comment by shashwatahalder01** [comment]
Scenario:
Popup stills open for add new product if vendor doesn't have any previous products


Screenshot:
Screenshot at Jul 19 10-46-03
Screenshot at Jul 19 10-46-17

**Comment by shashwatahalder01** [comment]
Scenario:
product gets published directly on dokan lite, even though publish directly is disabled for that vendor

1. add a vendor 
2. add a product for that vendor

Screenshot:
Screenshot at Jul 19 11-58-23

**Comment by shashwatahalder01** [comment]
Scenario:
wholesale setting doen't load on single page product create form

Steps to recreate:
1. enable wholesale from admin 
2. go to product create form
3. check wholesale checkbox

Screenshot:
Screenshot at Jul 19 12-48-16
Should be:
Screenshot at Jul 19 13-00-54

**Comment by shashwatahalder01** [comment]
Scenario:
 All  sub options regarding rma type loading at the same time on single page product create form. Should only load sub options based on RMA type

Steps to recreate:
1. Enable RMA from both Admin and Vendor
2.go to product create form
3.go to rma section [all options are loading]

Screenshot:

Screenshot at Jul 19 13-02-00

**Comment by shohag121** [comment]
@shashwatahalder01 Bhai, Issues are fixed.

**Comment by shohag121** [comment]
@Mohaiminulislam1989 Bhaya, Product creation page-related settings re-added.

Screenshot 2023-08-14 at 10 19 18 AM

**Pull request merged**